feat(GSoC 25): Brush tool

Abdelhadi-Wael opened the pull request:
Implementing the Brush tool as a part of Google Summer of Code 2025.

Abdelhadi-Wael committed:
feat(GSoC 25): Brush tool

Project Goals :

    Smooth Drawing and Erasing:
    The new overlay preview system shows strokes instantly as you draw or erase, using a fast offscreen buffer instead of updating the layer in real time. This avoids lag, improves stroke quality, and ensures smooth results. Final rendering happens only after the stroke ends.

    Reliable & Flexible Undo/Redo:
    Each stroke is a single undoable action. Users can choose between three undo modes—pixel buffer (faster, higher memory), re-render (lower memory, slower), or checkpointing (a mix of both)—depending on their performance and resource needs.

    Customizable Brushes: A new Brush Profile Editor will allow you to create, tweak (size, opacity, hardness, spacing, etc.), and manage your brush presets.

---

    Implemented the logic for creating and managing the overlay buffer for the Brush tool.
    When the mouse is pressed, a surface is created, and passed to the overlay renderer.
    The overlay renderer composites this buffer onto the canvas (to make it visible on the screen)

---

    Reused the logic for brush configuration and parsing settings from the existing state_brush.
    Integrated the MyPaint brush library engine to render strokes into a temporary buffer.
    Modified the drawing logic to clamp brush dabs within the layer bounds, preventing the layer from expanding.

Now, it’s possible to load and use different. myb brush presets and draw with them into the buffer in real-time, though drawing to the final layer is not yet implemented.

note: some brushes are not compatible with the version of ‘libmypaint’ we are using.

---

you will be able to either :

1 - import an image to draw on

2 - just start using the brush and a new canvas size bitmap layer will be created to draw on

---

first goal is achieved 2 more to go : )

In these two weeks, I finished the Smooth Drawing and Erasing milestone. The overlay preview system is now fully functional.

also a new feature is added → create image layer which allows you to create a transparent image layer with your desired resolution

---

mplemented Undo Mode 1 (Surface Saving):
This mode works by saving the entire image data before each stroke and restoring it during undo. This means image data is stored for every single stroke.

Advantages:

    Very fast and reliable.

Disadvantages:

    High memory usage:
    Memory Usage = Number of strokes × image size

---

Implemented Redraw Mode
This is the second undo mode. Instead of saving the entire image for each stroke, this mode works by saving only the stroke data itself (points, used brush settings). When you undo, it clears the layer and redraws all the previous strokes from the saved data.

Advantages:

    Low memory usage: It only stores stroke data, not large bitmap data.

Disadvantages:

    Slower performance: Can become slow with a large number of strokes, as it needs to re-render everything on each undo action.

Implemented Checkpointing Mode
This third mode is a hybrid approach that balances memory usage and performance. It works by automatically saving a full snapshot of the image (a “checkpoint”) every ‘N’ strokes. To undo, it reverts to the last checkpoint and then redraws only the few strokes made since then.

Advantages:

    Balanced performance and memory: It’s faster than the Redraw mode and uses less memory than saving the surface after every single stroke.

---

mplemented the Brush Profile Editor, which lets you view and control brush settings such as opacity, radius, hardness, and more. This was the third and final project goal — so we’ve now achieved all the project goals :tada: :smile:.

---

Original commit history:

initial structure

add UI for selecting brush

overlay

Add overlay surface rendering to canvas

fix :ci build fail

feat(brush): add MyPaint brushlib integration to brush tool

- Integrate brushlib::Brush
- Add BrushConfig class for parsing .myb brush definition files
- Add  brush palette UI with icon loading and selection
- Add brush settings persistence and coordinate transformation
- Replace simple overlay test with real brush stroke rendering

fix(brush): clamp brush bounds to surface dimensions instead of expansion

refactor(renderer): optimize brush overlay rendering and make it zoom independent

feat(brush): add permanent rendering to layer

feat(brush): add permanent rendering to layer

fix(stroke): applied new stroke system

- Simplified coordinate transforms by clipping rendering area to canvas bounds
- Copied image layer to overlay at stroke start, hid image layer during stroke for smooth erasing

feat(brush): add image layer creation dialog

feat(brush): implement undo mode 1 (surface saving)

feat(brush): implement undo mode 1 (surface saving)

feat(ui): cursor brush preview, radius control, color control

fix(brush): black outline

fix : group layer , limit radius

feat(brush): add undo redraw mode

add checkpointing undo mode

add undo mode selection dialog

feat(brush) : add brush profile editor

refactor(brush): fix indentation

fix(brush) : undo issue

fix (brushes path) : synchronization between preferences and brush state

- Force update of App::brushes_path when paths are modified

This resolves the bug where changing brush paths in preferences while using
another tool would not be reflected when switching to brush tool.

style: fix code formatting

minor fixes

add image data saving

fix(brush) :apply transformations to cairo surface

improve brush settings tab reset button behavior and remove color change settings

fix(save path): ensure modified bitmaps are saved to absolute paths

previously the `save_layer` function would take the 'filename'
parameter from a modified layer and pass it directly, after saving this path is converted from absolute to relative causing problems.

allow saving embedded layers

feat(brush) : surface expansion

fix(brush): replace Rect with tl, br

The Rect constructor normalize its coordinates on creation. This was causing a bug where flips in the brush preview were being discarded, because the coordinate order that represented the flip was being lost.

fix(brush): black outline

refactor(brush): prevent canvas from expanding during erase

fix cursor/dab mismatch

add tooltip for brushes buttons

remove slow tracking from default brushes

set tl,br while expanding

remove the old brush tool and change the new brush tool icon

remove unused functions

fix: crash on blend composite

fix: double free crash on application quit

style: formatting and consistency improvements

style: formatting and consistency improvements

fix(brush): preserve per brush settings when switching between tools or brushes

style(brush): fix style and consistency improvements

mohamedAdhamc commented:
Hello @Abdelhadi-Wael  I'll check this in a day or two. Any idea why the build errors are happening ?

Abdelhadi-Wael commented:
> Hello @Abdelhadi-Wael I'll check this in a day or two. Any idea why the build errors are happening ?

sounds good , the ci build fail is now fixed

Abdelhadi-Wael commented:
Hey @mohamedadhamc, the overlay system is almost complete. Currently I’m facing two issues:

1 - Erasing : Erasing doesn’t take effect until mouse-up because we’re only applying changes to the actual layer after the stroke is finished. Do you think we should apply it directly to the layer during drawing like the old brush (slow)? or is there a better approach?

2 -Stroke quality : The stroke currently looks pixelated at the default resolution settings. Do you think we should enforce a minimum resolution to ensure acceptable quality?

morevnaproject commented:
Hey @Abdelhadi-Wael 
Thank you for sharing your progress! I have tested your changes - looks very promising!

Regarding your questions:

> 1 - Erasing : Erasing doesn’t take effect until mouse-up because we’re only applying changes to the actual layer after the stroke is finished. Do you think we should apply it directly to the layer during drawing like the old brush (slow)? or is there a better approach?

I understand the problem with erasing you have.
Instead of using (slow) approach of old brush, I suggest the following:
- when user starts erasing, copy contents of active  image layer to temporary surface (overlay) and hide the active image layer.
- when user makes a stroke you modify the temporary overlay surface.
- when user finishes drawing, place contents of temporary surface to image layer and unhide it.
Approach is the same as you use during regular brush stroke.

> 2 -Stroke quality : The stroke currently looks pixelated at the default resolution settings. Do you think we should enforce a minimum resolution to ensure acceptable quality?

Let's leave it as is for now. 
Brush creates image with same resolution as current canvas - that's okay.
Later we probably will  change default canvas resolution for new documents.

As additional feature, it would be nice to add a "Create image layer" button in Brush Tool Options Panel.
<img width="1679" height="579" alt="image" src="https://github.com/user-attachments/assets/e73e4965-a780-4274-87da-5141a365899a" />
When user clicks the button it will show dialog where user can set parameters for new image layer:
- width
- height
The dialog will also have two buttons: "Create" and "Cancel"
When user clicks  "Create" button in this dialog then the new image layer created with given parameters.

Abdelhadi-Wael commented:
> As additional feature, it would be nice to add a "Create image layer" button in Brush Tool Options Panel. <img 
> * width
> * height
>   The dialog will also have two buttons: "Create" and "Cancel"
>   When user clicks  "Create" button in this dialog then the new image layer created with given parameters.

this feat is now implemented.

<img width="1857" height="1017" alt="Screenshot from 2025-07-15 12-16-03" src="https://github.com/user-attachments/assets/5433f0ea-146a-495e-a1c3-787887144232" />

Abdelhadi-Wael commented:
surface saving undo mode is ready : )

mohamedAdhamc commented:
Hello @Abdelhadi-Wael, I was testing this a bit more now. So far it is looking, great progress so far. 

One thing I noticed though that would be quite helpful to be implemented in the erase mode is as follows.

This is the current behaviour of the erase mode:
[Erasing.webm](https://github.com/user-attachments/assets/cdc6194a-b0e8-4bec-aefa-69b587fd21fc)

As you can see one doesn't have much fine control over erasing. So I suggest that while in erase mode switch the cursor to a circular looking one and only erase in the radius of that "cursor circle" and also have its radius be adjustable. I think this would greatly improve usability of the tool.

Abdelhadi-Wael commented:
thanks for the feedback 

> in erase mode switch the cursor to a circular looking one and only erase in the radius of that "cursor circle" and also 
have its radius be adjustable. I think this would greatly improve usability of the tool.

great idea i think it would be useful for drawing mode too, i will implement it and push asap.

morevnaproject commented:
Hi! Thanks for the fixes!
I have noticed an issue if you choose bright color for drawing - there is a black outline around the stroke.
Here is the video - 

https://github.com/user-attachments/assets/f05f09d1-4791-4a63-aa25-20beada03540

Also, I have noticed there is a delay/distance from cursor while drawing -

https://github.com/user-attachments/assets/d203b3ac-a09f-4246-a601-b7d2b855d4ae

Abdelhadi-Wael commented:
> I have noticed an issue if you choose bright color for drawing - there is a black outline around the stroke.

i will look into this 

> Also, I have noticed there is a delay/distance from cursor while drawing -

i noticed that too and i found out that it is not a bug.
it happens only with concept design brushes i tried other brushes like 
[deevad4-mypaint-brushkit.zip](https://github.com/user-attachments/files/21480949/deevad4-mypaint-brushkit.zip) and there was no delay

i also tried to paint with the concept design brushes in mypiant and same delay happens so it is probably something in the brush settings.

https://github.com/user-attachments/assets/990dd279-06d5-4716-8473-e3bf2862da2d

mohamedAdhamc commented:
Hello,

I tested out the two recent additions/fixes briefly and they seem to be good. Will test more thoroughly later, I noticed an issue related to the brush size button that wasn't introduced by this PR that I'll fix myself.

Abdelhadi-Wael commented:
hey @morevnaproject @mohamedAdhamc 

sorry for the late reply 
the issues are now resolved.

Abdelhadi-Wael commented:
all 3 undo modes are ready. please test them and let me know if you encounter any issues.

Abdelhadi-Wael commented:
hi @mohamedAdhamc @morevnaproject
I’ve added controls for all 42 brush settings.  what do you think?

1. Do you think we should keep all the settings or remove some? For example, there are 9 color related settings  should we remove them since we already have a color panel? Also, some options look like duplicates (e.g. “Dabs per basic radius” vs “Dabs per actual radius”).

2. What extra features would you like to see added to the brush settings editor ?

rodolforg commented on synfig-studio/src/gui/workarearenderer/renderer_brush_overlay.cpp:
```diff
@@ -0,0 +1,177 @@
+
+    cairo_surface->flush();
+
+    union { int i; char c[4]; } checker = {0x01020304};
+    bool big_endian = checker.c[0] == 1;
```
@ice0 @mohamedAdhamc maybe we should follow the algorithm to detect endianess suggested here:
 https://stackoverflow.com/questions/1001307/detecting-endianness-programmatically-in-a-c-program
instead of using `union` tricks.

Perhaps create it as a method in synfig/os.h?

Copy-and-pasted code:
```c++
enum class endianness
{
    little = 0,
    big = 1,
};

inline endianness get_system_endianness()
{
    const int value { 0x01 };
    const void * address { static_cast<const void *>(&value) };
    const unsigned char * least_significant_address { static_cast<const unsigned char *>(address) };

    return (*least_significant_address == 0x01) ? endianness::little : endianness::big;
}
```

rodolforg commented on synfig-studio/src/synfigapp/actions/layerbrush.h:
```diff
@@ -0,0 +1,131 @@
+**	\brief Template File
+**
+**	\legal
+**	......... ... 2014 Ivan Mahonin
```
Missing correct copyright line?

rodolforg commented on synfig-studio/src/synfigapp/actions/layerbrush.h:
```diff
@@ -0,0 +1,131 @@
+
+/* === S T A R T =========================================================== */
+
+#ifndef __SYNFIG_APP_ACTION_LAYERBRUSH_H
+#define __SYNFIG_APP_ACTION_LAYERBRUSH_H
```
remove the initial double underscore `__`. We are avoiding using it in new files, as it is reserved in C/C++

References:
https://stackoverflow.com/questions/43461615/c-double-underscore-inside-of-define-macro
https://devblogs.microsoft.com/oldnewthing/20230109-00/?p=107685
https://github.com/protobuf-c/protobuf-c/issues/188

rodolforg commented on synfig-studio/src/synfigapp/actions/layerbrush.cpp:
```diff
@@ -0,0 +1,474 @@
+**	\brief Template File
+**
+**	\legal
+**	......... ... 2014 Ivan Mahonin
```
missing correct copyright line.
Something like:
**	Copyright (c) 2025      Synfig Contributors

rodolforg commented on synfig-studio/src/synfigapp/actions/layerbrush.cpp:
```diff
@@ -0,0 +1,474 @@
+
+using namespace synfig;
+using namespace synfigapp;
+using namespace Action;
```
Is this one really necessary?

mohamedAdhamc commented:
Hello @Abdelhadi-Wael,

I noticed a bug in the undo functionality:

[undo-bug.webm](https://github.com/user-attachments/assets/f7a31d1a-63ec-49f9-aaaf-3f26acaa50d9)

Abdelhadi-Wael commented on synfig-studio/src/gui/states/state_brush2.cpp:
```diff
@@ -0,0 +1,1382 @@
+	layer_bounds = transform_stack_.empty() ? layer_bounds : transform_stack_.perform(layer_bounds);
+	overlay_rect_ = layer_bounds;
+
+	// copy layer to overlay
```
this does not handle the case where image is rotated any idea how should i fix this ?

https://github.com/user-attachments/assets/55d5dbc6-1065-498a-a778-f067e0e1922d

morevnaproject commented:
4. 

Next step is to implement image saving.
Here is how it should work:

- User imports an image file.
- Paints on image file using brush tool.
- User chooses "File" - "Save" and Synfig should save main file and all modified images. Currently it doesn't saves modified image:
![Peek 2025-09-03 10-40](https://github.com/user-attachments/assets/c5e0e346-56ad-4c61-869f-5d4cd3c2cfee)

morevnaproject commented:
5. 
 
One more thing remaining is to bring back automatic expanding of canvas boundaries -
<img width="809" height="252" alt="image" src="https://github.com/user-attachments/assets/fbf4d7b3-aad1-4566-a46f-ddfab8fe45a4" />

Abdelhadi-Wael committed:
fix(brush): prevent canvas from expanding during erase

Abdelhadi-Wael committed:
fix(brush): avoid black outline in brush strokes

"I have noticed an issue if you choose bright color for drawing - there
is a black outline around the stroke."

Reported-by: Konstantin Dimitriev
Reported-via: https://github.com/synfig/synfig/pull/3605#issuecomment-3130646185

mohamedAdhamc commented:
@Abdelhadi-Wael 

Ok, so I just did some very basic testing. The outline, and rotation issues seem to be fixed. However there are two things I noticed.

1.  The "eraser brush" always seems to have a different size than the brush size showing in the canvas. I'm not sure if this is something related to this tool or just related to the brush itself.
2. I think it would be helpful to have some sort of tool-tip if possible showing when hovering over the brushes. 

That's all that came to attention for now I'll be back tomorrow for some more testing and I'll be going over the code as well.

Abdelhadi-Wael committed:
refactor: remove slow tracking from default brushes

mohamedAdhamc commented:
> > strokes become more pixelated on release than the way they look while stroking.
> 
> @mohamedAdhamc The strokes appear more pixelated on release because you’re currently using preview rendering mode which reduces the quality. In final rendering mode the strokes will look identical to how they appear while drawing.

Ok, yes makes sense.

zaynopenapp reviewed (approved):
Is there still a problem? Can it be built?

mohamedAdhamc commented:
Hello guys, I was moving to another country 2 days ago.  So it was taking some time to organize things from my end. I should be able to continue reviewing soon but so far things were looking good :).

rodolforg commented:
I'm ok about removing the old broken tool.
The new one does what the current was supposed to do and more.

rodolforg commented:
@morevnaproject

rodolforg commented:
@mohamedAdhamc , I can use it normally, this issue doesn't happen to me.
I can start a new layer directly with new Brush tool, draw, save the .sifz, close it and load it back.
I tried to import a .png image, edit it with the switch layer selected and it worked too. Even with two "sublayers" it handled properly. Saved and load it back too.

@Abdelhadi-Wael However, Synfig crashes sometimes on closing o.O I couldn't identify how to reproduce yet though.

Abdelhadi-Wael commented:
@mohamedAdhamc I don’t know why this issue happens on your system; I think it’s something environment-specific rather than a problem in the code.

> @Abdelhadi-Wael However, Synfig crashes sometimes on closing o.O I couldn't identify how to reproduce yet though.

fixed in the latest commit (dc38b7b)

mohamedAdhamc commented:
> I think it’s something environment-specific rather than a problem in the code.

It only happens when using the brush tool though. For example loading a regular image and saving works fine. Sama as with all other layer types. So I wonder what the issue is.

mohamedAdhamc commented on synfig-studio/src/gui/dialogs/dialog_setup.cpp:
```diff
@@ -258,6 +258,18 @@ Dialog_Setup::create_system_page(PageInfo pi)
+	attach_label_section(pi.grid, _("Brush Undo Mode"), ++row);
+	pi.grid->attach(brush_undo_mode_combo, 1, row, 1, 1);
+	brush_undo_mode_combo.set_hexpand(true);
+	brush_undo_mode_combo.append("0", _("Redraw (low memory , slower)"));
```
```suggestion
	brush_undo_mode_combo.append("0", _("Redraw (low memory, slower)"));
```

mohamedAdhamc commented on synfig-studio/src/gui/dialogs/dialog_setup.cpp:
```suggestion

```

This seems to no longer be need. Unused var.

mohamedAdhamc commented:
Hello, slight code review. I will continue again tomorrow/the day after.

Abdelhadi-Wael commented:
> slight code review

Thanks for the review . I have made the changes.

mohamedAdhamc commented on synfig-studio/src/gui/states/state_brush2.cpp:
```diff
@@ -0,0 +1,1449 @@
+	Canvas::Handle canvas,
+	Layer::Handle layer,
+	CanvasView::Handle canvas_view,
+	TransformStack& transform_stack);
```
```suggestion
	static bool build_transform_stack(Canvas::Handle canvas, Layer::Handle layer, CanvasView::Handle canvas_view, TransformStack& transform_stack);
```

style

mohamedAdhamc commented on synfig-studio/src/gui/states/state_brush2.cpp:
```diff
@@ -0,0 +1,1449 @@
+	out_value.clear();
+	const char *p = *pos;
+	while ((*p >= 'a' && *p <= 'z') || *p == '_') ++p;
+	if (p > *pos) { out_value.assign(*pos, p); *pos = p; return true; }
```
style (same as before)

mohamedAdhamc commented on synfig-studio/src/gui/states/state_brush2.cpp:
```diff
@@ -0,0 +1,1449 @@
+	const char *num_start = p;
+	while(*p >= '0' && *p <= '9')
+		out_value = 10.f*out_value + (float)(*(p++) - '0');
+	if (p <= num_start) return false;
```
style (if and statement on separate lines)

mohamedAdhamc commented on synfig-studio/src/gui/states/state_brush2.cpp:
```diff
@@ -0,0 +1,1449 @@
+			out_value += (amplifier *= 0.1f)*(float)(*(p++) - '0');
+	}
+
+	if (negative) out_value = -out_value;
```
style

mohamedAdhamc commented on synfig-studio/src/gui/states/state_brush2.cpp:
```diff
@@ -0,0 +1,1449 @@
+				MapEntry entry;
+				const char *pp = p;
+				while(read_space(&pp) && (out_value.mapping.empty() || (read_key(&pp, ",") && read_space(&pp))) && read_map_entry(&pp, entry))
+					{ out_value.mapping.push_back(entry); p = pp; }
```
style

mohamedAdhamc commented on synfig-studio/src/gui/states/state_brush2.cpp:
```diff
@@ -0,0 +1,1449 @@
+	{
+		Glib::RefPtr<Gio::File> file = Gio::File::create_for_path(filename);
+		goffset s = file->query_info()->get_size();
+		if (s < 0) return;
```
style

mohamedAdhamc commented on synfig-studio/src/gui/states/state_brush2.cpp:
```diff
@@ -0,0 +1,1449 @@
+StateBrush2_Context::BrushConfig::apply(brushlib::Brush &brush)
+{
+	// Apply all loaded settings and input mappings to the brush instance
+	for(int i = 0; i < BRUSH_SETTINGS_COUNT; ++i)
```
```suggestion
	for (int i = 0; i < BRUSH_SETTINGS_COUNT; ++i)
```

mohamedAdhamc commented on synfig-studio/src/gui/states/state_brush2.cpp:
```diff
@@ -0,0 +1,1449 @@
+	for(int i = 0; i < BRUSH_SETTINGS_COUNT; ++i)
+	{
+		brush.set_base_value(i, settings[i].base);
+		for(int j = 0; j < INPUT_COUNT; ++j)
```
```suggestion
		for (int j = 0; j < INPUT_COUNT; ++j)
```

mohamedAdhamc commented on synfig-studio/src/gui/states/state_brush2.cpp:
```diff
@@ -0,0 +1,1449 @@
+		for(std::vector<InputEntry>::const_iterator j = settings[i].inputs.begin(); j != settings[i].inputs.end(); ++j)
+		{
+			brush.set_mapping_n(i, j->input, (int)j->mapping.size());
+			for(std::vector<MapEntry>::const_iterator k = j->mapping.begin(); k != j->mapping.end(); ++k)
```
```suggestion
			for (std::vector<MapEntry>::const_iterator k = j->mapping.begin(); k != j->mapping.end(); ++k)
```

mohamedAdhamc commented on synfig-studio/src/gui/states/state_brush2.cpp:
```diff
@@ -0,0 +1,1449 @@
+		{
+			String value;
+			App::brushes_path.clear();
+			//int count = atoi(value.c_str());
```
```suggestion

```

Abdelhadi-Wael commented:
@mohamedAdhamc sorry for the delay, I’ve made the changes.

mohamedAdhamc reviewed (commented):
Mostly, what is left now is to review `layerbrush`. I will try to do so shortly. Otherwise looks good thusfar.

mohamedAdhamc reviewed (dismissed):
Looks good to me :). Apart from the issue I was having about not being able to save on my system. But apparently, no-one else seems to experience it, so I am assuming perhaps it is unrelated.

rodolforg commented on synfig-core/src/synfig/surface.h:
```diff
@@ -59,7 +59,7 @@ class ColorPrep
 	static Color uncook_static(Color x)
 	{
-		if(!x.get_a())
+		if(fabs(x.get_a()) < 1e-8)
```
you can use `synfig::approximate_zero()`

rodolforg commented:
> I am not sure what is the root cause of this. But I also tried to look a little more to hopefully make it easier for you to try and figure it out.


@mohamedAdhamc ,
Can you get (via debugger) or print the value of `tmpfile` and `target::name__` (both on the mentioned `Instance::save_surface()`)?
Or temporarily set a callback to pass to `target->add_frame(&surface, nullptr);` instead of this `nullptr`? It would let it print the error in console.

rodolforg commented on synfig-studio/src/gui/states/state_brush2.cpp:
```diff
@@ -0,0 +1,1527 @@
+	void load_settings();
+	void save_settings();
+	bool scan_directory(const String &path, int scan_sub_levels, std::set<String> &out_files);
+	void select_brush(Gtk::ToggleToolButton *button, String filename);
```
code style: `&` and `*` should be next to Type not to variable name

rodolforg commented on synfig-studio/src/gui/states/state_brush2.cpp:
```diff
@@ -0,0 +1,1527 @@
+	void clear_overlay_preview();
+
+	void create_brushes_tab(Gtk::Notebook *notebook);
+	void create_settings_tab(Gtk::Notebook *notebook);
```
code style: `&` and `*` should be next to Type not to variable name

rodolforg commented on synfig-studio/src/gui/states/state_brush2.cpp:
```diff
@@ -0,0 +1,1527 @@
+	void create_brushes_tab(Gtk::Notebook *notebook);
+	void create_settings_tab(Gtk::Notebook *notebook);
+
+	void draw_to(Vector pos, Real pressure);
```
maybe `const Vector&` ?

rodolforg commented:
> > I am not sure what is the root cause of this. But I also tried to look a little more to hopefully make it easier for you to try and figure it out.
> 
> @mohamedAdhamc , about your comment [#3605 (comment)](https://github.com/synfig/synfig/pull/3605#issuecomment-3402736608):
> 
> Can you get (via debugger) or print the value of `tmpfile` and `target::name__` (both on the mentioned `Instance::save_surface()`)? Or temporarily set a callback to pass to `target->add_frame(&surface, nullptr);` instead of this `nullptr`? It would let it print the error in console.

@mohamedAdhamc ping :)

mohamedAdhamc commented:
Hello,

So currently for some reason I can't even use the brush tool anymore. It keeps saying brush stroke failed.

The following are the various things that come up:

Error on save:
<img width="723" height="331" alt="error_on_save" src="https://github.com/user-attachments/assets/c7733bcf-69b2-40b2-b284-417d571615a6" />

Various logs:
<img width="1060" height="173" alt="Screenshot from 2026-02-16 19-22-38" src="https://github.com/user-attachments/assets/0c5b2cfa-ce0a-4de0-829e-5d1070bcbbbb" />

rodolforg commented:
@mohamedAdhamc  is there a user-writable `/tmp` folder? :P

rodolforg commented:
@mohamedAdhamc  ping XD

mohamedAdhamc commented:
> @mohamedAdhamc ping XD

Hello,

Just rechecked, yes. Also, it is working now.

rodolforg commented:
I'll rebase this PR, squash and merge.
@mohamedAdhamc @Svarov-RZM @ice0 @morevnaproject Should I delete the old state_brush and related stuff? Rename the `state_brush2` to `state_brush`?

mohamedAdhamc commented:
> I'll rebase this PR, squash and merge. @mohamedAdhamc @Svarov-RZM @ice0 @morevnaproject Should I delete the old state_brush and related stuff? Rename the `state_brush2` to `state_brush`?

Yes, I think that sounds good. :)

rodolforg commented:
I have the original PR here in my machine. I rebased to current master branch, smashed all the commits, and rewrite the commit message with all of his reports on Synfig forums.

rodolforg commented:
hmm.. I think we'll redo it.
Keep the brushlib changes with its detailed commits.

rodolforg commented:
> hmm.. I think we'll redo it. Keep the brushlib changes with its detailed commits.

Done. Way better :)

rodolforg committed:
feat: add hotkey 'h' for new Brush tool

rodolforg commented on synfig-studio/src/synfigapp/actions/layerbrush.cpp:
```diff
@@ -0,0 +1,533 @@
+};
+
+static std::vector<StrokeData> strokes_history;
+static std::map<Layer_Bitmap::Handle, Surface> original_layer_surface;
```
@mohamedAdhamc Hmm... I think this static variables in Action will break stuff when someone has two synfig docs open and mixes Brush tool usages between them!

rodolforg replied:
Maybe it should be one for each root canvas.

mohamedAdhamc replied:
Good catch, yes I think a static variable in there sounds like a bad idea for when there are multiple canvases open. I wonder why it was made static to begin with though. And yes, I think one for each root canvas sounds reasonable.

rodolforg replied:
I think it is done to be able to replay all strokes but the 'last' one on undoing.

rodolforg replied:
@mohamedAdhamc Done.

rodolforg committed:
refactor: reinforce constness

rodolforg committed:
fix: avoid memory leak on error

rodolforg committed:
refactor: prefer filesystem::Path

rodolforg commented on synfig-studio/src/gui/states/state_brush2.cpp:
```diff
@@ -0,0 +1,1538 @@
+StateBrush2 studio::state_brush2;
+
+// brush -> (idx -> val)
+static std::map<String, std::map<int, float>> custom_settings;
```
Here is another `static`. I don't know if it is intentional to share between open files or not.

mohamedAdhamc replied:
I think here it might be intentional, to have the settings carry over.

rodolforg commented:
@mohamedAdhamc @Svarov-RZM @morevnaproject the only thing now preventing me to merge this PR is:
- Should I remove the old (and completely broken) brush tool (files `state_brush.h`/`cpp` and `layerpaint.h`/`cpp`)?
  - If so, should all those changes here be applied to `state_brush.cpp` and not `state_brush2.cpp`?

mohamedAdhamc commented:
> @mohamedAdhamc @Svarov-RZM @morevnaproject the only thing now preventing me to merge this PR is:
> 
> * Should I remove the old (and completely broken) brush tool (files `state_brush.h`/`cpp` and `layerpaint.h`/`cpp`)?
>   
>   * If so, should all those changes here be applied to `state_brush.cpp` and not `state_brush2.cpp`?

Hello, I think if history would be kept intact then already having it be just one version without the 2 (the new one), sounds like a good idea.

morevnaproject commented:
> @mohamedAdhamc @Svarov-RZM @morevnaproject the only thing now preventing me to merge this PR is:
> 
> - Should I remove the old (and completely broken) brush tool (files `state_brush.h`/`cpp` and `layerpaint.h`/`cpp`)?
> 
>   - If so, should all those changes here be applied to `state_brush.cpp` and not `state_brush2.cpp`?

Let's remove the old tool 👍
